src/dc_uudecode.c

```diff
@@ -32,31 +86,564 @@
  *     If no uuencoded parts are found or on errors, NULL is returned.
  */
 char* dc_uudecode_do(const char* text, char** ret_binary, size_t* ret_binary_bytes, char** ret_filename)
 {
 	// CAVE: This function may be called in a loop until it returns NULL, so make sure not to create an invinitive look.
 
-	if (text == NULL || ret_binary == NULL || ret_binary_bytes == NULL || ret_filename == NULL) {
+	if( text == NULL || ret_binary == NULL || ret_binary_bytes == NULL || ret_filename == NULL ) {
 		goto cleanup; // bad parameters
 	}
 
+	char* uustartpos    = NULL;
+	char* ret_msg_text  = NULL; // NULL = no uuencoded parts are found (standard case)
+	
+	// first make a quick check if a part exists
+	uustartpos = dc_find_uuencoded_part(text);
+	if (!uustartpos){
+		// no or no further uuencoded parts found
+		goto cleanup;
+	}
+
+	/**
+	 * optimization possible: identify first call of this function to unify buffer only once
+	 */
+	// prepare buffer and unify buffer
+	char* txt = strdup(text);
+	dc_unify_lineends(txt);
+
+	// find again in new buffer
+	uustartpos = dc_find_uuencoded_part(txt);
+	if (uustartpos){
+		// then handle uuencoded part
+		ret_msg_text = dc_handle_uuencoded_part (txt, uustartpos, ret_binary, ret_binary_bytes, ret_filename);
+	}
+	free(txt);
+	
+	return ret_msg_text;
+	
+
 cleanup:
+
 	return NULL;
 }
 
 
+
+/***********************************************************************
+ *  Helper functions
+ */
+
+int dc_make_error_txt(char** ret_binary, const char* format, int nargs, int n1, int n2, int n3){
+	/**
+	 * Format error text and
+	 *  realloc ret_binay to necessary size and
+	 *  copy error text to ret_binary
+	 * 
+	 *  @return - len of error txt
+	 */
+	 
+	int 	ret_bytes = 0;
+	char 	msg[strlen(format)+33]; // for 3 parameters
+	
+	switch(nargs){
+		case 0: strcpy(msg, format); break;
+		case 1: sprintf(msg, format, n1); break;
+		case 2: sprintf(msg, format, n1, n2); break;
+		case 3: sprintf(msg, format, n1, n2, n3); break;
+		default:
+			return ret_bytes;
+	}
+
+	int      l = strlen(msg);
+	*ret_binary = (char*) realloc(*ret_binary, l+1);
+	
+	if (*ret_binary){
+		// write error msg to output file if possible!
+		strncpy(*ret_binary, msg, l);
+		ret_bytes = l * (-1);
+	}
+	else{
+		ret_bytes = 0;
+	}
+	
+	return ret_bytes;
+}
+
+
+char* dc_print_hex (char* s){
+	/**
+	 * make hex representation of a string
+	 * 
+	 * @Return
+	 *      hex representation of s (needs to be free'd)
+	 * 
+	 */
+	int n = strlen(s);
+	int i;
+	char* hex = (char*)calloc((2*n+1), sizeof(char));
+	
+	for(i = 0; i<n; i++){
+		sprintf(hex+i*2, "%02X", s[i]);
+	}
+	
+	return hex;
+	
+} //dc_print_hex()
+
+
+
+int dc_getline (char** ret_lineptr, char* source, char** ret_nextcharptr){
+	/**
+	 * 	Extracts first line from a given string (source)
+	 * 
+	 * 	Instead of strtok_r it doesn't manipulate the input string 
+	 *
+	 *  Parameters:
+	 * 
+	 *  @ret_lineptr   - returned line incl. line end char
```

Comment:
i think doxygen won't accept this format, it should read `@param ret_lineprt Explain line pointer`

Reply:
Ok, now I understand how this is used. Will update the function descriptions.

Reply:
Also, move outside of function.